draft-ietf-taps-impl.md

```diff
@@ -887,29 +887,82 @@ Connection Object:
+To map Connection objects to SCTP streams without head-of-line blocking on the sender
+side, both the sending and receiving SCTP implementation must support message interleaving {{!RFC8260}}.
+Both SCTP implementations must also support stream reconfiguration. Finally, both communicating endpoints
+must be aware of this intended multiplexing; {{NEAT-flow-mapping}} describes a
+way for a Transport System to negotiate the stream mapping capability using SCTP's adaptation layer indication,
+such that this functionality would only take effect if both ends sides are aware of it.
+The first flow, for which the SCTP association has been created, will always use stream id zero.
+All additional flows are assigned to unused stream ids in growing order. To avoid a conflict
+when both endpoints map new flows simultaneously, the peer which initiated the transport connection
+will use even stream numbers whereas the remote side will map its flows to odd stream numbers.
+Both sides maintain a status map of the assigned stream numbers. Generally, new streams
+must consume the lowest available (even or odd, depending on the side) stream number; this
+rule is relevant when lower numbers become available because Connection objects associated
+to the streams are closed.
+
+Initiate:
+: If this is the only Connection object that is assigned to the SCTP association or stream mapping has
+not been negotiated, CONNECT.SCTP is called. Else, a new stream is used: if there are enough streams
+available, `Initiate` is just a local operation that assigns a new stream number to the Connection object.
+The number of streams is negotiated as a parameter of the prior CONNECT.SCTP call, and it represents a
+trade-off between local resource usage and the number of Connection objects that can be mapped
+without requiring a reconfiguration signal. When running out of streams, ADD_STREAM.SCTP must be called.
+
+InitiateWithSend:
+: If this is the only Connection object that is assigned to the SCTP association or stream mapping has
+not been negotiated, CONNECT.SCTP is called with the "user message" parameter. Else, a new stream
+is used (see `Initiate` for how to handle running out of streams), and this just sends the first message
+on a new stream.
+
+Ready:
+: `Initiate` or `InitiateWithSend` returns without an error, i.e. SCTP's four-way handshake has completed. If an association with the peer already exists, and stream mapping has been negotiated and enough streams are available, a Connection Object instantly becomes Ready after calling `Initiate` or `InitiateWithSend`.
+
+InitiateError:
+: Failure of CONNECT.SCTP.
+
+ConnectionError:
+: TIMEOUT.SCTP or ABORT-EVENT.SCTP.
+
+Listen:
+: LISTEN.SCTP. If an association with the peer already exists and stream mapping has been negotiated, `Listen` just expects to receive a new message on a new stream id (chosen in accordance with the stream number assignment procedure described above) and will return upon reception of this message.
```

Comment:
Wouldn't this make the Listen call blocking? Shouldn't the Listener object be returned immediately and then ConnectionReceived be called once the message arrives?

Reply:
Good catch, thanks! Fixed. There was no need to say "will return upon reception".

JFYI: anyway these primitives are abstract - just a writing style to find the original part of the spec for the functionality. E.g., the text describing "LISTEN.SCTP" in section 4 of RFC 8303 says: "Pass 1 primitive/event: 'Initialize', followed by the 'Communication Up' or 'Restart' notification and possibly the  'Adaptation Layer' notification" - so in fact it's (supposed to be) non-blocking even with the standard SCTP socket interface, under the hood.